cpp/include/raft/sparse/matrix/preprocessing.cuh

```diff
@@ -0,0 +1,568 @@
+  float* results,
+  bool bm25_on,
+  float k_param,
+  float b_param)
+{
+  cudaStream_t stream = raft::resource::get_cuda_stream(handle);
+  auto nnz            = csr_in.structure_view().get_nnz();
+  auto rows           = raft::make_device_vector<IndexType, int64_t>(handle, nnz);
+  auto columns        = raft::make_device_vector<IndexType, int64_t>(handle, nnz);
+  auto values         = raft::make_device_vector<ValueType, int64_t>(handle, nnz);
+  raft::sparse::matrix::detail::convert_csr_to_coo(
+    handle, csr_in, rows.view(), columns.view(), values.view());
+  transform(
+    handle, rows.view(), columns.view(), values.view(), nnz, results, bm25_on, k_param, b_param);
+}
+
+/**
+ * This function transforms the csr matrix based on statistics collected during fit
+ * cycle.
+ *
+ * @tparam ValueType
+ *   Type of the values in the sparse matrix.
+ * @tparam IndexType
+ *   Type of the indices associated with the values.
+ *
+ * @param[in] handle
+ *   Container for managing reusable resources.
+ * @param[in] rows
+ *   array coo representation of rows for non zero values.
+ * @param[in] columns
+ *   array coo representation of columns for non zero values.
+ * @param[in] values
+ *   array coo representation of non zero values.
+ * @param[in] nnz
+ *   The number of non-zero values in the coo array representations.
+ * @param[in] results
+ *   array that is the size of the csr_in nnz that will store the encoded
+ *      transform values.
+ * @param[in] bm25_on
+ *   When true calculates bm25 encoding values, otherwise tfidf.
+ * @param[in] k_param
+ *   bm25 okapi optimization parameter k1
+ * @param[in] b_param
+ *   bm25 okapi optimization parameter b
+ * */
+template <typename ValueType, typename IndexType>
+void SparseEncoder<ValueType, IndexType>::transform(
+  raft::resources const& handle,
+  raft::device_vector_view<IndexType, int64_t> rows,
+  raft::device_vector_view<IndexType, int64_t> columns,
+  raft::device_vector_view<ValueType, int64_t> values,
+  IndexType nnz,
+  ValueType* results,
+  bool bm25_on,
+  float k_param,
+  float b_param)
+{
+  cudaStream_t stream = raft::resource::get_cuda_stream(handle);
+
+  int maxLimit   = nnz;
+  int blockSize  = (maxLimit < 128) ? maxLimit : 128;
+  int num_blocks = (maxLimit + blockSize - 1) / blockSize;
+  float avgIdLen = (ValueType)fullIdLen / numRows;
+  int* counts;
+  cudaMallocManaged(&counts, maxLimit * sizeof(IndexType));
+  cudaMemset(counts, 0, maxLimit * sizeof(IndexType));
+  raft::sparse::op::coo_sort(IndexType(rows.size()),
+                             IndexType(columns.size()),
+                             IndexType(values.size()),
+                             rows.data_handle(),
+                             columns.data_handle(),
+                             values.data_handle(),
+                             stream);
+  raft::sparse::matrix::detail::_scan<<<num_blocks, blockSize>>>(rows.data_handle(), nnz, counts);
```

Comment:
Since we have a `stream`, we'd better to apply it in the kernels
```suggestion
  raft::sparse::matrix::detail::_scan<<<num_blocks, blockSize, 0, stream>>>(rows.data_handle(), nnz, counts);
```